[groovy] strange error message when using anonymous groovy objects

<!-- Please read our Code of Conduct: https://github.com/deephaven/deephaven-core/blob/main/CODE_OF_CONDUCT.md -->
<!-- Please search existing issues to avoid creating duplicates. -->

**Description**

A clear and concise description of what the bug is.

**Steps to reproduce**
```groovy
x = new Object() {
    long get(long ii) { return ii; }
}
y = emptyTable(1).update("X = x[ii]")
```

**Expected results**

An error message of the following style:
```
Having trouble with the following expression:
Full expression           : x[ii]
Expression having trouble : io.deephaven.db.tables.lang.DBLanguageParser$VisitArgs@1e199514
Exception message         : Cannot find method get(long) in interface groovy.lang.GroovyObject
```

**Actual results**

```
/io/deephaven/temp/c415813373987228486v61_0/Formula.java:80: error: <identifier> expected
    private final io.deephaven.dynamic.Script_6.1 x;
                                               ^
/io/deephaven/temp/c415813373987228486v61_0/Formula.java:80: error: <identifier> expected
    private final io.deephaven.dynamic.Script_6.1 x;
                                                   ^
/io/deephaven/temp/c415813373987228486v61_0/Formula.java:89: error: ')' expected
        x = (io.deephaven.dynamic.Script_6.1) __params[0].getValue();
```

**Additional details and attachments**

I suspect it's a JDK upgrade side effect.

I have a fix but interferes with the PyObject #1490 - so will put up the the PR immediately when that merges.

--- a/engine/table/src/main/java/io/deephaven/engine/table/impl/select/QueryScopeParamTypeUtil.java
+++ b/engine/table/src/main/java/io/deephaven/engine/table/impl/select/QueryScopeParamTypeUtil.java
@@ -1,6 +1,7 @@
 package io.deephaven.engine.table.impl.select;
 
 import groovy.lang.Closure;
+import groovy.lang.GroovyObject;
 import io.deephaven.compilertools.CompilerTools;
 import io.deephaven.util.type.TypeUtils;
 
@@ -96,6 +97,8 @@ public class QueryScopeParamTypeUtil {
             type = ((Enum<?>) value).getDeclaringClass();
         } else if (value instanceof Closure) {
             type = Closure.class;
+        } else if (value instanceof GroovyObject) {
+            type = GroovyObject.class;
         } else {
             type = value.getClass();
         }

--- a/engine/table/src/test/java/io/deephaven/engine/util/scripts/TestGroovyDeephavenSession.java
+++ b/engine/table/src/test/java/io/deephaven/engine/util/scripts/TestGroovyDeephavenSession.java
@@ -7,6 +7,7 @@ package io.deephaven.engine.util.scripts;
 import io.deephaven.base.verify.Assert;
 import io.deephaven.engine.table.Table;
 import io.deephaven.engine.table.TableDefinition;
+import io.deephaven.engine.table.impl.select.FormulaCompilationException;
 import io.deephaven.engine.util.GroovyDeephavenSession;
 import io.deephaven.engine.liveness.LivenessScope;
 import io.deephaven.engine.liveness.LivenessScopeStack;
@@ -57,6 +58,21 @@ public class TestGroovyDeephavenSession {
         final TableDefinition definition = y.getDefinition();
         final Class<?> colClass = definition.getColumn("X").getDataType();
         Assert.equals(colClass, "colClass", java.util.List.class);
+    }
+
+    @Test
+    public void testAnonymousObject() {
+        final String script = "x = new Object() {\n" +
+                "  long get(long ii) { return ii; }\n" +
+                "}\n" +
+                "y = emptyTable(1).update(\"X = x[ii]\")";
+        try {
+            session.evaluateScript(script);
+        } catch (FormulaCompilationException exception) {
+            Assert.eqTrue(exception.getCause().getCause().getMessage().contains(
+                    "Cannot find method get(long) in interface groovy.lang.GroovyObject"),
+                    "exception contains helpful error message");
+        }
     }
 }
 

